src/compiler/checker.ts

```diff
@@ -32138,7 +32140,8 @@ namespace ts {
             }
 
             const functionFlags = getFunctionFlags(func);
-            const type = returnType && unwrapReturnType(returnType, functionFlags);
+            const unwrappedType = returnType && unwrapReturnType(returnType, functionFlags);
+            const type = unwrappedType && getReducedType(unwrappedType);

```

Comment:
I think this is best covered in a separate PR since it is really an unrelated fix.